docs(graphql): add federation 2 doc (#2461)
nestjs/docs.nestjs.com: content/graphql/federation.md

})
export class AppModule {}
```

### Federation 2

To quote the [Apollo docs](https://www.apollographql.com/docs/federation/federation-2/new-in-federation-2), Federation 2 improves developer experience from the original Apollo Federation (called Federation 1 in this doc), which is backward compatible with most original supergraphs.

> warning **Warning** Mercurius doesn't fully support Federation 2. You can see the list of libraries that support Federation 2 [here](https://www.apollographql.com/docs/federation/supported-subgraphs#javascript--typescript).

In the following sections, we'll upgrade the previous example to Federation 2.

#### Federated example: Users

> One change in Federation 2 is that entities has no originating subgraph, so we don't need to extend `Query` anymore. For more detail please refer to [this topic](https://www.apollographql.com/docs/federation/federation-2/new-in-federation-2#entities) in Apollo docs.

Reviewer: ```suggestion
One change in Federation 2 is that entities have no originating subgraph, so we don't need to extend `Query` anymore. For more detail please refer to [the entities topic](https://www.apollographql.com/docs/federation/federation-2/new-in-federation-2#entities) in Apollo Federation 2 docs.
```

#### Schema first

We can simply remove `extend` keyword from the schema.

```graphql
type User @key(fields: "id") {
  id: ID!
  name: String!
}

type Query {
  getUser(id: ID!): User
}
```

#### Code first

> To use Federation 2, we need to specify in `autoSchemaFile` option.

Author: I don't see anywhere in the Apollo doc mention it as "version 2", but since it follows Apollo releasing stages I think we can assume it's technically a major version 2.

import { UsersService } from './users.service'; // Not included in this example

@Module({
  imports: [
    GraphQLModule.forRoot<ApolloFederationDriverConfig>({
      driver: ApolloFederationDriver,
      autoSchemaFile: {
        federation: 2,
      },
    }),
  ],
  providers: [UsersResolver, UsersService],
})
export class AppModule {}
```

#### Federated example: Posts

With the same reason as above, we don't need to extend `User` and `Query` anymore.

#### Schema first

We can simply remove `extend` and `external` directives from the schema

```graphql
type Post @key(fields: "id") {
  id: ID!
  title: String!
  body: String!
  user: User
}

type User @key(fields: "id") {
  id: ID!
  posts: [Post]
}

type Query {
  getPosts: [Post]
}
```

#### Code first

Since we don't extend `User` entity anymore, we can simply remove `extends` and `external` directives from `User`.

```ts
import { Directive, ObjectType, Field, ID } from '@nestjs/graphql';
import { Post } from './post.entity';

@ObjectType()
@Directive('@key(fields: "id")')
export class User {
  @Field((type) => ID)
  id: number;

  @Field((type) => [Post])
  posts?: Post[];
}
```

> Also, similar to User service, we need to specify `GraphQLModule` to use Federation 2.

Reviewer: ```suggestion
Also, similarly to the User service, we need to specify in the `GraphQLModule` to use Federation 2.
```
